source/MRMesh/MRRadiusCompensation.cpp

```diff
@@ -0,0 +1,269 @@
+#include "MRRadiusCompensation.h"
+#include "MRDistanceMapParams.h"
+#include "MRDistanceMap.h"
+#include "MRParallelFor.h"
+#include "MRMesh.h"
+#include "MRMakeSphereMesh.h"
+#include "MRRegionBoundary.h"
+#include "MRMeshComponents.h"
+#include "MRPositionVertsSmoothly.h"
+#include "MRBitSetParallelFor.h"
+#include "MR2to3.h"
+#include "MRExpandShrink.h"
+#include "MRMeshRelax.h"
+#include "MRMeshDelone.h"
+#include "MRMeshDecimate.h"
+
+namespace MR
+{
+
+Expected<void> compensateRadius( Mesh& mesh, const CompensateRadiusParams& params )
+{
+    auto sb0 = subprogress( params.callback, 0.0f, 0.7f );
+    const auto& faceRegion = mesh.topology.getFaceIds( params.region );
+    VertBitSet vertRegion = getIncidentVerts( mesh.topology, faceRegion );
+
+    if ( MeshComponents::hasFullySelectedComponent( mesh, vertRegion - mesh.topology.findBoundaryVerts( &vertRegion ) ) )
+        return unexpected( "MeshPart should not contain closed components" );
+
+    MeshPart mp = MeshPart( mesh, params.region );
+
+    MeshToDistanceMapParams dmParams;
+    if ( params.pixelSize > 0 )
+        dmParams = MeshToDistanceMapParams( params.direction, Vector2f::diagonal( params.pixelSize ), mp, true );
+    else
+        dmParams = MeshToDistanceMapParams( params.direction, Vector2i::diagonal( 200 ), mp, true );
+
+    if ( !reportProgress( sb0, 0.1f ) )
+        return unexpectedOperationCanceled();
+    
+    DistanceMapToWorld wParams = DistanceMapToWorld( dmParams );
+
+    auto dm = computeDistanceMap( mp, dmParams, subprogress( sb0, 0.1f, 0.2f ) );
+    if ( dm.size() == 0 )
+        return unexpectedOperationCanceled();
+
+    Vector2f realPixelSize;
+    realPixelSize.x = dmParams.xRange.length() / dmParams.resolution.x;
+    realPixelSize.y = dmParams.yRange.length() / dmParams.resolution.y;
+
+    Vector2i pixelsInDiameter;
+    pixelsInDiameter.x = int( std::ceil( 2 * params.toolRadius / realPixelSize.x ) ) + 1;
+    pixelsInDiameter.y = int( std::ceil( 2 * params.toolRadius / realPixelSize.y ) ) + 1;
+
+    constexpr std::array<Vector2i, 9> cNeigborsOrder =
+    {
+        Vector2i( -1,1 ),
+        Vector2i( 0,1 ),
+        Vector2i( 1,1 ),
+        Vector2i( 1,0 ),
+        Vector2i( 1,-1 ),
+        Vector2i( 0,-1 ),
+        Vector2i( -1,-1 ),
+        Vector2i( -1,0 ),
+        Vector2i( -1,1 )
```

Comment:
why do we repeat first as last?

Reply:
for easier iterating and taking into account last pixel too (we take `i` and `i+1` on each iteration)